src/main/java/com/onarandombox/MultiverseCore/commandtools/flags/FlagsManager.java

```diff
@@ -0,0 +1,172 @@
+ */
+public class FlagsManager {
+    private final Map<String, FlagGroup> flagGroupMap;
+
+    /**
+     * Creates a new FlagsManager.
+     */
+    public FlagsManager() {
+        flagGroupMap = new HashMap<>();
+    }
+
+    /**
+     * Registers a flag group.
+     *
+     * @param flagGroup The target flag group to register.
+     */
+    public void registerFlagGroup(@NotNull FlagGroup flagGroup) {
+        flagGroupMap.put(flagGroup.getName(), flagGroup);
+    }
+
+    /**
+     *  Get a flag group by name.
+     *
+     * @param groupName The target flag group name.
+     * @return The flag group if found, null otherwise.
+     */
+    public @Nullable FlagGroup getFlagGroup(@Nullable String groupName) {
+        return this.flagGroupMap.get(groupName);
+    }
+
+    /**
+     * Autocompletes suggestions for flags.
+     *
+     * @param groupName The target flag group name.
+     * @param flags The current flags so far.
+     * @return The list of suggestions.
+     */
+    public @NotNull Collection<String> suggest(@Nullable String groupName, @NotNull String[] flags) {
+        FlagGroup flagGroup = this.getFlagGroup(groupName);
+        if (flagGroup == null) {
+            return Collections.emptyList();
+        }
+
+        Collection<String> suggestions = new ArrayList<>();
+        MVFlag currentFlag = (flags.length <= 1) ? null : flagGroup.getFlagByKey(flags[flags.length - 2]);
+
+        if (currentFlag instanceof MVValueFlag) {
+            MVValueFlag valueFlag = (MVValueFlag) currentFlag;
+            if (valueFlag.getCompletion() != null) {
+                suggestions.addAll((Collection<String>) valueFlag.getCompletion().get());
+            }
+            if (valueFlag.isOptional()) {
+                suggestions.addAll(getRemainingFlagKeys(flagGroup, flags));
+            }
+        } else {
+            suggestions.addAll(getRemainingFlagKeys(flagGroup, flags));
+        }
+
+        return suggestions;
+    }
+
+    /**
+     * Gets the flags keys that has not been used yet.
+     *
+     * @param flagGroup The target flag group.
+     * @param flags The current flags so far.
+     * @return The remaining flag keys.
+     */
+    private @NotNull Set<String> getRemainingFlagKeys(@NotNull FlagGroup flagGroup, @NotNull String[] flags) {
+        Set<String> keysRemaining = new HashSet<>(flagGroup.getKeys());
+        for (String flag : flags) {
+            MVFlag mvFlag = flagGroup.getFlagByKey(flag);
+            if (mvFlag != null) {
+                keysRemaining.remove(mvFlag.getKey());
+            }
+        }
+        return keysRemaining;
+    }
+
+    /**
+     * Parses the flags.
+     *
+     * @param groupName The target flag group name.
+     * @param flags The flags to parse.
+     * @return The parsed flags.
+     *
+     * @throws InvalidCommandArgument If the flags are invalid.
+     */
+    public @NotNull ParsedFlags parse(@Nullable String groupName, @NotNull String[] flags) {
```

Comment:
Any way to break this method down and make it smaller? Lot of complexity in this thing. Maybe a separate `FlagParser` is needed to hold the operating state.

Reply:
yea the logic is quite confusing, but works :laughing: